cpu/cc2538/syscalls.c

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2014, Institute for Pervasive Computing, ETH Zurich.
+ * All rights reserved.
+ *
+ * Author: Andreas Dröscher <contiki@anticat.ch>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the Institute nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include <sys/stat.h>
```

Comment:
If I take `_open()`, I see it defined in `cpu/arm/at91sam7s/newlib-syscalls.c`, `cpu/stm32w108/e_stdio/src/syscalls.c` and `cpu/stm32w108/hal/micro/cortexm3/stm32w108/crt-stm32w108.c`. They're indeed dummy, which means that they probably had to be introduced only because the build system is broken (see the archive group build trick just discussed, and try to put `-lc` at various places without the archive group, and ld will eventually complain about `_open()` and the like not being defined). Newlib does not provide these.

If dummy implementations are needed, something has to be wrong somewhere, because it means that an unsupported feature of Newlib is used, or that the build system is broken. If something unsupported is used, then we want to detect it at build time, so using dummy implementations to fix the build process is wrong. By not doing this on CC2538, the link will fail if something unsupported is used, which is what we want.

Apart from that, there are many things duplicated on many platforms, including only on ARM platforms, not only to implement common features, but also because of diverging local APIs for the same features. All common stuff should be factored out and reused as much as possible, but doing this from the current state would be a huge work.